### PR TITLE
New build System

### PR DESCRIPTION
**Pull request by mgautierfr**
This branch is mainly a cleanup/rebase of #20 

- With far less commits
- Less unnecessary files added and removed
- No adaption of wrapper itself (so it doesn't compile with current version)

While it "compile" (with the change in PR #23) it is NOT TESTED.
The testing part itself doesn't work (and I'm not sure how it should work, @MohitMaliFtechiz we have to discuss about this).

@MohitMaliFtechiz please review this PR and check that I haven't forget something compared to #20.
There is few questions to you in the commits' messages.

There is still few things I disagree with this PR but I suggest we merge it (if there is nothing obviously wrong) as it is and fix the issues later, one by one.

### CODE REVIEW

File: .github/workflows/pull_request.yml [hunk 1]
```diff
@@ -0,0 +1,46 @@
...
+        with:
+          fetch-depth: 1
+
+      - name: Set up JDK 11
+        uses: actions/setup-java@v1
+        with:
+          java-version: 11
+
+      - name: Build Project
+        run: bash contrib/build.sh
```
Comment (MohitMaliFtechiz) [inline .github/workflows/pull_request.yml:25]: it should be ```bash install_deps.sh```

Comment (MohitMaliFtechiz) [inline .github/workflows/pull_request.yml:25]: it should be ```bash install_deps.sh```

File: .github/workflows/pull_request.yml [hunk 2]
```diff
@@ -0,0 +1,46 @@
...
+
+      - name: Compile Project
+        run: |
```
Comment (MohitMaliFtechiz) [inline .github/workflows/pull_request.yml:28]: For new folder structure we don't need to go into android-libkiwixbuild we should remove this. now we can directly use ```./gradlew androidSourcesJar build```

File: .github/workflows/pull_request.yml [hunk 3]
```diff
@@ -0,0 +1,46 @@
...
+          cd android-libkiwixbuild
+          ./gradlew androidSourcesJar build
+
+      - name: create unit coverage
+        run: |
```
Comment (MohitMaliFtechiz) [inline .github/workflows/pull_request.yml:33]: Here too.

File: .github/workflows/release.yml [hunk 4]
```diff
@@ -0,0 +1,49 @@
...
+          fetch-depth: 1
+
+      - name: Set up JDK 11
+        uses: actions/setup-java@v2
+        with:
+          distribution: adopt
+          java-version: 11
+
+      - name: Building Project
+        run: bash contrib/build.sh
```
Comment (MohitMaliFtechiz) [inline .github/workflows/release.yml:29]: Here we should use ```bash install_deps.sh```

File: .github/workflows/release.yml [hunk 5]
```diff
@@ -0,0 +1,49 @@
...
+
+      - name: Release build
+        run: |
```
Comment (MohitMaliFtechiz) [inline .github/workflows/release.yml:32]: For new folder structure we don't need to go into android-libkiwixbuild we should remove this. Now we can directly use ```./gradlew androidSourcesJar build```

File: .github/workflows/release.yml [hunk 6]
```diff
@@ -0,0 +1,49 @@
...
+          cd android-libkiwixbuild
+          ./gradlew assemble
+
+      - name: Genrate Source jar
+        run: |
```
Comment (MohitMaliFtechiz) [inline .github/workflows/release.yml:37]: Here too.

File: lib/build.gradle [hunk 7]
```diff
@@ -0,0 +1,289 @@
...
+
+task checkCurrentJavaVersion() {
+    if (JavaVersion.current() != JavaVersion.VERSION_11) {
+        throw new RuntimeException("This build must be run with java 11. your current java version is ${JavaVersion.current()}")
+    }
+}
+
+
+task generateHeaderFilesFromJavaWrapper(type: Exec) {
+    workingDir "${projectDir}/src/java/org/kiwix/"
```
Comment (MohitMaliFtechiz) [inline lib/build.gradle:287]: It should be ```workingDir"${projectDir}/src/main/java/org/kiwix/"```

**Comment by MohitMaliFtechiz** [comment]
hi @mgautierfr ,

Related to testing part , Test cases were written purely in java they are not android specific test cases , so to run test cases written on java required .so files which are compiled on Linux variant not android variant so that why we are compiling its separately.

**Comment by mgautierfr** [comment]
I've indeed forget to adapt the github workflow to the new file structure. Thanks. It should be ok now.

> Related to testing part , Test cases were written purely in java they are not android specific test cases , so to run test cases written on java required .so files which are compiled on Linux variant not android variant so that why we are compiling its separately.

I don't think that testing part should compile the library to test. We should have a target to build the library  (and different targets for different archs) and having the test target depending of the target building the library for linux x86_64

**Comment by MohitMaliFtechiz** [comment]
> I've indeed forget to adapt the github workflow to the new file structure. Thanks. It should be ok now.
> 
> > Related to testing part , Test cases were written purely in java they are not android specific test cases , so to run test cases written on java required .so files which are compiled on Linux variant not android variant so that why we are compiling its separately.
> 
> I don't think that testing part should compile the library to test. We should have a target to build the library (and different targets for different archs) and having the test target depending of the target building the library for linux x86_64

We are only building wrapper files to generate .so file while for libzim and libkiwiz we are downloading linux variant .so file from download.kiwix.org

**Comment by mgautierfr** [comment]
Yes. Here what we should have (big picture):
- 5 "targets" downloading the libzim.so for the 5 archs (4 android and 1 linux native)
- 5 "targets" downloading the libkiwix.so for the 5 archs (as libzim)
- 1 "target" generating the c headers of the wrapper from java classes (maybe 2 targets if we handle wrappers around libzim and libkiwix separately)
- 5 "targets" creating the wrappers .so for the 5 archs. (Maybe 10 targets if we handle wrappers separately). The targets should depends of the downloading target and header generation target.
- 1 "target" to build the aar from all .so. This target depends of the 4/8 "creating wrapper" targets (android only)
- 1 test target which depends of the linux native "creating wraper" and "just" use the .so to test everything is ok

With the chain of dependencies. We would only have to launch `./gradlew build` or `./gradlew test` to have everything downloaded and compiled.

**Comment by mgautierfr** [comment]
I propose to merge this PR as it is and fix testing and target dependencies later.
@MohitMaliFtechiz, is it ok for you ?

**Comment by MohitMaliFtechiz** [comment]
> I propose to merge this PR as it is and fix testing and target dependencies later.
> @MohitMaliFtechiz, is it ok for you ?

Yes its fine for now.

**Pull request merged**